RooFit::MultiProcess & TestStatistics part 1: mathcore & Minuit2 changes

egpbos opened the pull request:
This PR is the first part of a split and clean-up of #8294, as suggested by @guitargeek.

In this PR, only the changes to mathcore and Minuit2 are taken into account. These changes will be necessary for the later RooFit parts which will come in separate PRs.

Specifically, this PR introduces the ExternalInternalGradientCalculator in Minuit2. This subclass of AnalyticalGradientCalculator enables calculation of the derivative outside of Minuit2, but in "Minuit-internal" parameter space. This allows exact replication of the gradients as calculated in Minuit2 itself, which in turn enables external optimization and parallelization of this process without having to modify Minuit2, without having to worry about differences in outcome. In other words, the resulting external gradient calculation can be easily unit tested against the existing Minuit2 gradient calculation.

~Note: still working on "cherry picking" all changes and organizing them into functionally sensible commits. Will un-draft the PR once this is done.~ Ready for review.

guitargeek commented:
Thank you so much for investing the time to make new commits!

bellenot commented:
@egpbos please fork also [rootest](https://github.com/root-project/roottest) to prevent timeouts on the Windows build nodes, thanks in advance.

egpbos commented:
@bellenot thanks for the tip. Should I fork to the same org where this ROOT dev branch is at (github.com/roofit-dev) or to my personal account?

guitargeek commented:
Just a fork to your personal account is sufficient. The scrips are looking for `egpbos/roottest`. You don't need to add any commits to this fork of roottest, just forking.

egpbos commented:
@phsft-bot build just on windows10/cxx14

Axel-Naumann commented on math/mathcore/inc/Math/IFunction.h:
```diff
@@ -232,6 +236,29 @@ namespace ROOT {
+       a separate subclass that can do second derivatives (and step sizes, see DoStepSize), but this
+       was chosen because of easier integration into the existing Minuit2 / Fitter framework.
+      */
+      virtual T DoSecondDerivative(const T */*x*/, unsigned int /*icoord*/) const {
```
```suggestion
      virtual T DoSecondDerivative(const T * /*x*/, unsigned int /*icoord*/) const {
```

lmoneta commented on math/minuit2/inc/Minuit2/FCNGradientBase.h:
```diff
@@ -38,7 +43,24 @@ class FCNGradientBase : public FCNBase {
+   virtual std::vector<double> G2ndDerivative(const std::vector<double> &) const = 0;
+
+   virtual std::vector<double> GStepSize(const std::vector<double> &) const = 0;
+
```
same comment as before for the Capitalisation of the member functions

egpbos replied:
Resolved by removing.

lmoneta commented on math/minuit2/inc/Minuit2/MnUserTransformation.h:
```diff
@@ -91,6 +92,8 @@ class MnUserTransformation {
    // Index = internal Parameter
    double DInt2Ext(unsigned int, double) const;
+   double D2Int2Ext(unsigned int, double) const;
+   double GStepInt2Ext(unsigned int, double) const;
```
I think the GStepInt2Ext function is not needed and could be removed.
Se comment below

egpbos replied:
Indeed :)

lmoneta commented on math/minuit2/src/MnSeedGenerator.cxx:
```diff
@@ -160,13 +159,13 @@ MinimumSeed MnSeedGenerator::operator()(const MnFcn &fcn, const AnalyticalGradie
    if (ng2ls.HasNegativeG2(dgrad, prec)) {
-      Numerical2PGradientCalculator ngc(fcn, st.Trafo(), stra);
-      state = ng2ls(fcn, state, ngc, prec);
+      state = ng2ls(fcn, state, gc, prec);
```
I am not sure that removing the usage of the Numerical2PGradientCalculator is ok for all cases.
The following function , NegativeG2LineSearch::operator(), uses the G2 and GStep. If we have an analytical gradient not implementing these it can be problematic. 
Actually why not leaving it always in all cases ?

egpbos replied:
I have put it back in and it seems to have no effect on my tests, thanks for the suggestion.

lmoneta commented on math/minuit2/inc/Minuit2/FCNGradientBase.h:
```diff
@@ -38,7 +43,24 @@ class FCNGradientBase : public FCNBase {
 
    virtual std::vector<double> Gradient(const std::vector<double> &) const = 0;
 
+   virtual std::vector<double> G2ndDerivative(const std::vector<double> &) const = 0;
```
Do we need them at pure virtual functions ? I would add them as an option, returning a dummy vector in the base class.

egpbos replied:
We don't need them at all! ;) Removed, resolved.

lmoneta reviewed (changes requested):
Thank you @egpbos for this PR. 

I have some doubts about adding in the `Math/IFunction` interfaces the second derivatives and step sizes. I think, when supplying an external gradient to Minuit2, the G2 is used only when computing the Seed state and eventually used in NegativeG2LineSearch if G2<0.
Recently I have changed the entry point of the minimisation in ModularFunctionMinimizer, see 
https://github.com/root-project/root/blob/master/math/minuit2/src/ModularFunctionMinimizer.cxx#L142
when using a FCN computing the gradient. 
There the Numerical MInuit2 gradient is used for the seed, so there is really no needs to provide to Minuit G2 and step size's at the cost of using for the first iteration the Minuit2 gradient calculator. The cost of doing this I think is minimal given the fact that a large number of iteration at the end will be done ( at least 10 but often something like 100). 
Furthermore the MnHesse is using still the Numerical gradient, so I would prefer if for the time being we make the smallest changes and add in a later time full support for providing second derivatives, that can be used also in MnHesse.

egpbos commented on math/mathcore/inc/Math/IFunction.h:
```diff
@@ -360,6 +387,31 @@ namespace ROOT {
+          gstep[icoord] = BaseGrad::StepSize(x, icoord);
+        }
+      }
+
```
These functions were no longer necessary.

egpbos commented on math/minuit2/src/AnalyticalGradientCalculator.cxx:
```diff
@@ -44,21 +45,45 @@ FunctionGradient AnalyticalGradientCalculator::operator()(const MinimumParameter
+         if (fTransformation.Parameter(ext).HasLimits()) {
+            double int2ext_g2 = fTransformation.D2Int2Ext(i, par.Vec()(i));
+            double int2ext_gstep = fTransformation.GStepInt2Ext(i, par.Vec()(i));
+            vg2(i) = int2ext_g2 * g2[ext];
```
These functions were no longer necessary.

egpbos commented on math/minuit2/src/AnalyticalGradientCalculator.cxx:
```diff
@@ -44,21 +45,45 @@ FunctionGradient AnalyticalGradientCalculator::operator()(const MinimumParameter
+            double int2ext_g2 = fTransformation.D2Int2Ext(i, par.Vec()(i));
+            double int2ext_gstep = fTransformation.GStepInt2Ext(i, par.Vec()(i));
+            vg2(i) = int2ext_g2 * g2[ext];
+            vgstep(i) = int2ext_gstep * gstep[ext];
```
It's removed now.

egpbos commented:
@lmoneta I have removed all the second derivative and step size code. As you suspected, it was all unnecessary. Looking back at my logs, it seems I just forgot to clean them up after introducing ExternalInternalGradientCalculator. This, together with the `long double` precision upgrade of the parameter transformation classes, allows for perfect bitwise floating point replication of the gradient calculation outside of Minuit. Thanks for the thorough review!

I also took care of the copyright lines and fixed some additional layout issues.

Ready for a final review!

egpbos commented:
The JupyROOT failure is a master branch failure unrelated to this PR. Other than that, CI passes now. I think it should be ready to merge if you have no further change requests @lmoneta.

lmoneta commented on math/mathcore/inc/Math/IFunction.h:
```diff
@@ -360,7 +360,7 @@ namespace ROOT {
          }
 
-
+         virtual bool returnsInMinuit2ParameterSpace() const { return false; }
```
Is this function really needed ? It should not be at the gradient calculator level to know if the internal parameter space is used

lmoneta replied:
Looking at the code, I would add maybe a new flag to the 
`ROOT::Math::Minimiser::SetFunction( gradFunc & f, bool= false)`
In this way the Minuit2Minimizer can set this flag in the FCNGradientBase class when  creating it from the function interface passed by the clients (e.g. RooMinimizer).
If you prefer, we can merge with this and I can do this change later, please let me know

egpbos replied:
Your suggestion makes sense, indeed it feels weird that a Minuit detail is in mathcore. In light of time constraints, I would prefer to keep it like this for now, if you don't mind too much. The problem is that returnsInMinuit2ParameterSpace is also used in other places right now, in particular in the `TestStatistics::MinuitFcnGrad` class that will be introduced in the next PR I will make.

lmoneta commented on math/minuit2/inc/Minuit2/ExternalInternalGradientCalculator.h:
```diff
@@ -0,0 +1,42 @@
+
+/**********************************************************************
+ *                                                                    *
+ * Copyright (c) 2005 LCG ROOT Math team,  CERN/PH-SFT                *
```
Can you please remove the copyrights ?

lmoneta commented on math/minuit2/inc/Minuit2/ExternalInternalGradientCalculator.h:
```diff
@@ -0,0 +1,42 @@
+/**********************************************************************
+ *                                                                    *
+ * Copyright (c) 2005 LCG ROOT Math team,  CERN/PH-SFT                *
+ * Copyright (c) 2017 Patrick Bos, Netherlands eScience Center        *
```
Do you really want to keep this class name instead of calling something else ?

egpbos replied:
I'm open to suggestions, I don't love it either.

egpbos replied:
Maybe `InternalSpaceGradientCalculator`? Or `MinuitSpaceGradientCalculator`? Or `TransformedSpaceGradientCalculator`? I find it hard to make it more compact, while at the same time clearly indicating the relation to the other gradient calculators.

Perhaps as an alternative it would make sense to make a namespace `GradientCalculator` in which the different types could be gathered.

egpbos replied:
To at least make the use case clearer (and perhaps also the name), I added some doxygen documentation for the class. Hope this helps.

lmoneta commented on math/minuit2/inc/Minuit2/FCNGradAdapter.h:
```diff
@@ -56,7 +56,15 @@ class FCNGradAdapter : public FCNGradientBase {
-   bool CheckGradient() const { return false; }
+   bool CheckGradient() const override { return false; }
+
+   GradientParameterSpace gradParameterSpace() const override {
```
As mentioned above, I would add in this class the flag to distinguish external from internal gradient parameter space

lmoneta reviewed (changes requested):
Thank you for applying the changes I requested earlier. 
I have requested a change in Function, but I can do it later if you prefer.
For the ExternalInternalGradientCalculator class name, can we find a better name ? 
Otherwise we can discuss it later..

Axel-Naumann commented on math/minuit2/inc/Minuit2/SinParameterTransformation.h:
```diff
@@ -28,9 +28,9 @@ class SinParameterTransformation {
-   double Int2ext(double Value, double Upper, double Lower) const;
-   double Ext2int(double Value, double Upper, double Lower, const MnMachinePrecision &) const;
-   double DInt2Ext(double Value, double Upper, double Lower) const;
+   long double Int2ext(long double Value, long double Upper, long double Lower) const;
+   long double Ext2int(long double Value, long double Upper, long double Lower, const MnMachinePrecision &) const;
+   long double DInt2Ext(long double Value, long double Upper, long double Lower) const;
```
Should these be additional overloads, leaving the perf of the `double` version? Or have them as templates, only offering `double` and `long double` specializations (and possibly `float`)?

egpbos replied:
We discussed this, and concluded that it's unlikely to cause issues. The two potential problems would be:

1. On some more exotic platform the long doubles may not be implemented or differently and the (bit-wise exact) tests will not pass. However, it is unclear what platforms this would be. GPU computation, for instance, would not be affected, because the parameter transformations are not part of the main calculation.
2. This also means performance is not expected to be a big issue. With big fits, the parameter transformations are a minute part of the Minuit run time. Most time by far is spent in the function calls. The only situation in which it could be an issue is if you have a very cheap function with a huge amount of parameters. This seems quite unlikely for realistic fits.

egpbos replied:
One could also argue that a pro of using `long double` is that it clearly signals that these trigonometric transformations can lose some bits. This happens when you use the `double` versions and transform back and forth. In some situations, you can then lose 1 or 2 bits, just because of the trigonometric implementations. An alternative solution would be more precise sine etc functions.

Axel-Naumann replied:
Thanks!

> and the (bit-wise exact) tests will not pass

That's curious - how do you "survive" on Windows? https://docs.microsoft.com/en-us/cpp/cpp/fundamental-types-cpp?view=msvc-160#floating-point-types

lmoneta replied:
Windows will use the double version and it will then use few bits when using the external calculator with the multiprocess. Will the multiprocess be available on Windows ? Probably not

Axel-Naumann replied:
OK - but if there's only a `long double` version then which "double version" are you referring to? "Bit-wise exact" still depends on implementation / architecture. I trust this is handled, just making sure you're aware. (And I still don't know where these tests are.)

egpbos replied:
The tests in #8596 for instance @Axel-Naumann :) Btw, they don't pass currently either, due to changes in master that I hadn't incorporated yet. On the todo list.

But good question about Windows. I haven't tested at all there yet. So probably, at least for the time being, it will have to be disabled on Windows, as @lmoneta says. In principle, ZeroMQ works on all platforms, so that is not the bottleneck. However, I have no idea how/whether forking works on Windows.

egpbos replied:
Just so you don't have to dig into the code: the test compares the results of fitting with RooMinimizerFcn vs fitting the same function with RooGradMinimizerFcn. We put a lot of effort into making those two fits run *exactly* the same way. Small bit-wise differences quickly accumulate in the algorithm to end up with different results (past the desired precision, so e.g. in the 6th decimal if you wanted the fit to converge to 5).

lmoneta reviewed (approved):
I think we have clarified all issues and we are ready to merge this one.
Thank you Patrick for all this !
